storage/innobase/row/row0import.cc

```diff
@@ -4246,6 +4442,173 @@ fil_tablespace_iterate(
+		    "INSERT INTO SYS_FIELDS VALUES"
+		    "(:id, :pos, :name);\n"
+		    "END;\n", trx);
+
+  if (err != DB_SUCCESS)
+    my_error(ER_INTERNAL_ERROR, MYF(0),
+             "InnoDB: Insert into SYS_FIELDS failed");
+  return err;
+}
+
+/** Update the position of the columns and
+insert the hidden fts doc id in the sys columns table
+@param  table   table to be imported
+@param  fts_pos position of fts doc id column
+@param  trx     transaction
+@return DB_SUCCESS or error code */
+static
+dberr_t innodb_insert_hidden_fts_col(dict_table_t* table,
+                                     ulint  fts_pos,
+                                     trx_t* trx)
+{
+  dberr_t err= DB_SUCCESS;
+  pars_info_t*    info = nullptr;
+  for (ulint n= table->n_cols - 1; n >= fts_pos; n--)
+  {
+    info= pars_info_create();
+    pars_info_add_ull_literal(info, "id", table->id);
+    pars_info_add_ull_literal(info, "pos", n);
+    pars_info_add_ull_literal(info, "val", n + 1);
+    err= que_eval_sql(info,
+                    "PROCEDURE UPD_ALL_COL () IS\n"
+                    "BEGIN\n"
+                    "UPDATE SYS_COLUMNS SET\n"
+                    "POS = :val\n"
+                    "WHERE TABLE_ID = :id AND POS = :pos;\n"
+                    "END;\n", trx);
+    if (err != DB_SUCCESS)
+    {
+      my_error(ER_INTERNAL_ERROR, MYF(0),
+               "InnoDB: Updating SYS_COLUMNS failed");
+      return err;
+    }
+  }
+
+  info= pars_info_create();
+  dict_col_t* fts_col= &table->cols[fts_pos];
+  pars_info_add_ull_literal(info, "id", table->id);
+  pars_info_add_int4_literal(info, "pos", fts_pos);
+  pars_info_add_str_literal(info, "name", FTS_DOC_ID_COL_NAME);
+  pars_info_add_int4_literal(info, "mtype", fts_col->mtype);
+  pars_info_add_int4_literal(info, "prtype", fts_col->prtype);
+  pars_info_add_int4_literal(info, "len", fts_col->len);
+  pars_info_add_int4_literal(info, "base", 0);
+
+  err= que_eval_sql(info,
+                    "PROCEDURE ADD_COL () IS\n"
+                    "BEGIN\n"
+                    "INSERT INTO SYS_COLUMNS VALUES"
+                    "(:id,:pos,:name,:mtype,:prtype,:len,:base);\n"
+                    "END;\n", trx);
+  if (err != DB_SUCCESS)
+  {
+    my_error(ER_INTERNAL_ERROR, MYF(0),
+             "InnoDB: Insert into SYS_COLUMNS failed");
+    return err;
+  }
+
+  info= pars_info_create();
+  ulint n= dict_table_encode_n_col(
+    unsigned(table->n_cols) - DATA_N_SYS_COLS, table->n_v_cols)
+    | (table->flags & DICT_TF_COMPACT) << 31;
+
+  pars_info_add_int4_literal(info, "n", n);
+  pars_info_add_ull_literal(info, "id", table->id);
+
+  err= que_eval_sql(info,
+                    "PROCEDURE UPDATE_N_COLS () IS\n"
+                    "BEGIN\n"
+                    "UPDATE SYS_TABLES SET N_COLS = :n"
+                    " WHERE ID = :id;\n"
+                    "END;\n", trx);
+  if (err != DB_SUCCESS)
+    my_error(ER_INTERNAL_ERROR, MYF(0),
+             "InnoDB: Updating SYS_TABLES.N_COLS failed");
```

Comment:
I think that the user only needs to know that updating the data dictionary failed. Where exactly, it should not matter, because there hardly is anything that a user can do to correct such errors. Can we use just one procedure for everything? And add a `DBUG_EXECUTE_IF` to inject an error, so that we can check that no memory is leaked.

I checked that this transaction would be committed by `prebuilt->trx->commit()` in `row_import_cleanup()`. I did not find a call to a rollback anywhere. It would seem that in case of an error, `row_import_error()` would invoke `row_import_cleanup()`, which would wrongly commit any partially executed changes to the data dictionary tables.